cli/azd/pkg/azdo/pipeline.go

```diff
@@ -0,0 +1,231 @@
+		return nil, err
+	}
+
+	var exists bool = true
+	var count = 0
+	var maxTries = 4
+	for exists {
+		exists, err = pipelineExists(ctx, &client, &projectId, &name)
+		if err != nil {
+			return nil, err
+		}
+		count = count + 1
+
+		if exists {
+			name = fmt.Sprintf("%s - %s (%d)", name, repoName, count)
+		} else {
+			continue
+		}
+
+		if count >= maxTries {
+			return nil, fmt.Errorf("error creating new pipeline")
+		}
+	}
+
+	queue, err := getAgentQueue(ctx, projectId, connection)
+	if err != nil {
+		return nil, err
+	}
+
+	createDefinitionArgs, err := createAzureDevPipelineArgs(
+		ctx, projectId, name, repoName, credentials, env, queue, provisioningProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	newBuildDefinition, err := client.CreateDefinition(ctx, *createDefinitionArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	return newBuildDefinition, nil
+}
+
+// create Azure Deploy Pipeline parameters
+func createAzureDevPipelineArgs(
+	ctx context.Context,
+	projectId string,
+	name string,
+	repoName string,
+	credentials AzureServicePrincipalCredentials,
+	env environment.Environment,
+	queue *taskagent.TaskAgentQueue,
+	provisioningProvider provisioning.Options,
+) (*build.CreateDefinitionArgs, error) {
+
+	repoType := "tfsgit"
+	buildDefinitionType := build.DefinitionType("build")
+	definitionQueueStatus := build.DefinitionQueueStatus("enabled")
+	defaultBranch := fmt.Sprintf("refs/heads/%s", DefaultBranch)
+	buildRepository := &build.BuildRepository{
+		Type:          &repoType,
+		Name:          &repoName,
+		DefaultBranch: &defaultBranch,
+	}
+
+	process := make(map[string]interface{})
+	process["type"] = 2
+	process["yamlFilename"] = AzurePipelineYamlPath
+
+	variables := make(map[string]build.BuildDefinitionVariable)
+	variables["AZURE_SUBSCRIPTION_ID"] = createBuildDefinitionVariable(credentials.SubscriptionId, false, false)
+	if provisioningProvider.Provider == provisioning.Terraform {
+		variables["ARM_TENANT_ID"] = createBuildDefinitionVariable(credentials.TenantId, false, false)
+		variables["ARM_CLIENT_ID"] = createBuildDefinitionVariable(credentials.ClientId, true, false)
+		variables["ARM_CLIENT_SECRET"] = createBuildDefinitionVariable(credentials.ClientSecret, true, false)
+	}
+	variables["AZURE_LOCATION"] = createBuildDefinitionVariable(env.GetLocation(), false, false)
+	variables["AZURE_ENV_NAME"] = createBuildDefinitionVariable(env.GetEnvName(), false, false)
+	variables["AZURE_SERVICE_CONNECTION"] = createBuildDefinitionVariable(ServiceConnectionName, false, false)
+
+	agentPoolQueue := &build.AgentPoolQueue{
+		Id:   queue.Id,
+		Name: queue.Name,
+	}
+
+	trigger := make(map[string]interface{})
+	trigger["batchChanges"] = false
+	trigger["maxConcurrentBuildsPerBranch"] = 1
+	trigger["pollingInterval"] = 0
+	trigger["isSettingsSourceOptionSupported"] = true
+	trigger["defaultSettingsSourceType"] = 2
+	trigger["settingsSourceType"] = 2
+	trigger["defaultSettingsSourceType"] = 2
+	trigger["triggerType"] = 2
```

Comment:
@wbreza I'll make the change, and yes it would be nice if provided structs, they do for some parameter types but not all. A future contribution might be to add the missing structs to the golang sdk for azure devops.